scripts/xlinear_predict/predict.py

```diff
@@ -0,0 +1,311 @@
+"""
+This code is taken from the previous grants-tagger repository prior
+to removing support for xlinear. Specifically, at this commit:
+https://github.com/wellcometrust/grants_tagger/tree/ef1c065b92eec36d944c903a38a4c7921b93aaaa
+
+The code is used to load the xlinear model and predict MeSH terms for
+a given abstract. It is used in the scripts/create_xlinear_bertmesh_comparison_csv.py
+"""
+import pickle
+import logging
+import os
+import json
+import scipy.sparse as sp
+import numpy as np
+import ast
+
+from sklearn.feature_extraction.text import TfidfVectorizer
+from sklearn.base import BaseEstimator, ClassifierMixin
+
+from pecos.xmc.xlinear.model import XLinearModel
+from pecos.xmc import Indexer, LabelEmbeddingFactory
+
+logger = logging.getLogger(__name__)
+
+
+def load_pickle(obj_path):
+    with open(obj_path, "rb") as f:
+        return pickle.loads(f.read())
+
+
+def save_pickle(obj_path, obj):
+    with open(obj_path, "wb") as f:
+        f.write(pickle.dumps(obj))
+
+
+def format_predictions(
+    Y_pred_proba, label_binarizer, threshold=0.5, probabilities=True
+):
+    """
+    Formats predictions to output a list of dictionaries
+
+    Y_pred_proba: sparse array or list of predicted probabilites or class
+    (i.e. the output of  `.predict` or `.predict_proba` classifier)
+    label_binarizer: A sklearn fitted label binarizer
+    threshold: Float between 0 and 1
+    probabilities: Whether Y_pred_proba will contain probabilities or just predictions
+
+    Returns:
+        A list of dictionaries for each prediction, e.g.
+        [{"tag": "Malaria", "Probability": 0.5}, ...]
+    """
+    tags = []
+    for y_pred_proba in Y_pred_proba:
+        if sp.issparse(y_pred_proba):
+            y_pred_proba = np.asarray(y_pred_proba.todense()).ravel()
+        if probabilities:
+            tags_i = {
+                tag: prob
+                for tag, prob in zip(label_binarizer.classes_, y_pred_proba)
+                if prob >= threshold
+            }
+        else:
+            tags_i = [
+                tag
+                for tag, prob in zip(label_binarizer.classes_, y_pred_proba)
+                if prob >= threshold
+            ]
+        tags.append(tags_i)
+
+    return tags
+
+
+class MeshXLinear(BaseEstimator, ClassifierMixin):
```

Comment:
let's move this into models into its own folder

Reply:
You mean inside the `grants_tagger_light` package?

Reply:
yep, i know we said lets remove but since its used in the script lets place it in the right path, no need to use it anywhere else other than the script